app/src/main/java/de/danoeh/antennapod/activity/AddLocalFilesActivity.java

```diff
@@ -0,0 +1,170 @@
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
+        if (requestCode == PERMISSION_REQUEST_READ_EXTERNAL_STORAGE) {
+            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                chooseFileFromFilesystem(); // Retry
+            }
+        }
+    }
+
+    /*
+     * Lets the user choose a specific file to import.
+     */
+    private void chooseFileFromFilesystem() {
+        int permission = ActivityCompat.checkSelfPermission(this, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+        if (permission != PackageManager.PERMISSION_GRANTED) {
+            requestPermission();
+            return;
+        }
+
+        DialogProperties properties = new DialogProperties();
+        properties.selection_mode = DialogConfigs.SINGLE_MODE;
+        properties.selection_type = DialogConfigs.DIR_SELECT;
+        properties.root = new File(DialogConfigs.DEFAULT_DIR);
+        properties.error_dir = new File(DialogConfigs.DEFAULT_DIR);
+        properties.offset = new File(DialogConfigs.DEFAULT_DIR);
+        properties.extensions = null;
+
+        FilePickerDialog dialog = new FilePickerDialog(this, properties);
+        dialog.setTitle("Select a File");
+
+        dialog.setDialogSelectionListener(this);
+
+        dialog.show();
+    }
+
+    @Override
+    public void onSelectedFilePaths(String[] files) {
+        for (String f: files) {
+            Uri uri = new Uri.Builder().scheme("file").path(f).build();
+            Log.d(TAG, "path is " + uri.getPath());
+            boolean ret = importUri(uri);
+            if (ret == false) {
+                Toast.makeText(this, "Could not import", Toast.LENGTH_LONG).show();
+            } else {
+                Toast.makeText(this, "Success!", Toast.LENGTH_LONG).show();
+            }
+
+            //go back to main
+            finish();
+        }
+    }
+
+    private void requestPermission() {
+        String[] permissions = { android.Manifest.permission.READ_EXTERNAL_STORAGE };
+        ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_READ_EXTERNAL_STORAGE);
+    }
+
+    boolean importUri(@Nullable Uri uri) {
+        if (uri == null) {
+            new MaterialDialog.Builder(this)
+                    .content(R.string.opml_import_error_no_file)
+                    .positiveText(android.R.string.ok)
+                    .show();
+            return false;
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            int permission = ActivityCompat.checkSelfPermission(this, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+            if (permission != PackageManager.PERMISSION_GRANTED) {
+                return false;
+            }
+
+            try {
+                LocalFeedUpdater.startImport(uri, this);
```

Comment:
This is done on the ui thread. Therefore, if importing takes long, the app displays an "app not responding" error. I think it would be better to just create the feed (without items) here and then make the updater handle adding the actual files.

Reply:
How to trigger an update in another thread? It seems if I call any of the public `*refresh*Feed*` in `DBTasks` it will be done in the UI thread.

Reply:
I don't know either. Will have a look at this as soon as possible. Maybe we can find out how it is done when adding normal feeds